README.md
# Powder - encrypt passwords, store them in text safely

The extension allows to encrypt passwords with a master password.<br>
It enables CodeLens (inline text actions) for any text that looks like:

> <b>login:password</b>

Password part can be encrypted, decrypted and copied to clipboard.  

## Example

<sub><span style="text-decoration: underline;">Hide Password</span> | <span style="text-decoration: underline;">Copy Password</span></sub><br>
test<span>@example.<span>com:password123

↓ Click "Hide Password", enter master password ↓

<sub><span style="text-decoration: underline;">Unhide Password</span> | <span style="text-decoration: underline;">Copy Password</span></sub><br>
test<span>@example.<span>com:AAF5etlyDen/7nX6nbJ5IUY+v/1d5NC3RFMxLxnHhGwRfHOI4cfyQfWtSs41Z6OAdu/Gn78=

Unhide password leads to decryption.<br>
You can copy hidden password, master password is required.

## Encryption Details

Standard browser encryption API is used.<br>
AES-GCM 256 bit is used for the payload.<br>
Both the master password and the payload are salted with 196 bit of different random data.<br>

Encrypted data format is a base64 value of:
- 1 byte starting mark = 0
- 1 bytes version mark
- 12 bytes password salt
- 12 bytes payload salt
- N>0 byte payload, which contains:
  - length of data as a string
  - '\0' separator
  - payload

## Development

Contribution is welcome!

To start development:
- install `yarn`;
- clone the repository;
- run `yarn` command to install packages;
- press `F5` to test and debug;
- create PR after

## Roadmap

- Fix selection after encryption
- Handle errors on incorrect password
- Improve regex
- Encryption of all the lines in a file with one command
- Improving payload length handling
- More compact encrypted line without damaging encryption strength

Additionally:
- Associate with .pwd, make option to enable extension for these files only?
- Copy login?
- Should we also enable login encryption?
- What else can we encrypt?
- Powder mode - change Hide Password/Unhide Password with Powder/Unpowder

CHANGELOG.md
# Change Log

All notable changes to the "powder" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

webpack.config.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

//@ts-check
'use strict';

//@ts-check
/** @typedef {import('webpack').Configuration} WebpackConfig **/

const path = require('path');
const webpack = require('webpack');

/** @type WebpackConfig */
const webExtensionConfig = {
	mode: 'none', // this leaves the source code as close as possible to the original (when packaging we set this to 'production')
	target: 'webworker', // extensions run in a webworker context
	entry: {
		'extension': './src/web/extension.ts',
	},
	output: {
		filename: '[name].js',
		path: path.join(__dirname, './dist/web'),
		libraryTarget: 'commonjs',
		devtoolModuleFilenameTemplate: '../../[resource-path]'
	},
	resolve: {
		mainFields: ['browser', 'module', 'main'], // look for `browser` entry point in imported node modules
		extensions: ['.ts', '.js'], // support ts-files and js-files
		alias: {
			// provides alternate implementation for node module and source files
		},
		fallback: {
			// Webpack 5 no longer polyfills Node.js core modules automatically.
			// see https://webpack.js.org/configuration/resolve/#resolvefallback
			// for the list of Node.js core module polyfills.
			'assert': require.resolve('assert')
		}
	},
	module: {
		rules: [{
			test: /\.ts$/,
			exclude: /node_modules/,
			use: [{
				loader: 'ts-loader'
			}]
		}]
	},
	plugins: [
		new webpack.ProvidePlugin({
			process: 'process/browser', // provide a shim for the global `process` variable
		}),
	],
	externals: {
		'vscode': 'commonjs vscode', // ignored because it doesn't exist
	},
	performance: {
		hints: false
	},
	devtool: 'nosources-source-map', // create a source map that points to the original source file
	infrastructureLogging: {
		level: "log", // enables logging required for problem matchers
	},
};

module.exports = [ webExtensionConfig ];

src/web/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import {
	CancellationToken, CodeLens, CodeLensProvider, commands,
	env, EventEmitter, ExtensionContext, languages, Event,
	Position, Range, TextDocument, window, workspace
} from 'vscode';
import { isEncrypted, encrypt, decrypt } from './crypto-utils';

/**
 * CodelensProvider
 */
export class CodelensProvider implements CodeLensProvider {

	private codeLenses: CodeLens[] = [];
	private regex: RegExp;
	private _onDidChangeCodeLenses: EventEmitter<void> = new EventEmitter<void>();
	public readonly onDidChangeCodeLenses: Event<void> = this._onDidChangeCodeLenses.event;

	constructor() {
		this.regex = /[^\r\n:\s]+[:][^\r\n:\s]{3,}/g;

		workspace.onDidChangeConfiguration((_) => {
			this._onDidChangeCodeLenses.fire();
		});
	}

	public provideCodeLenses(document: TextDocument, token: CancellationToken): CodeLens[] | Thenable<CodeLens[]> {
		this.codeLenses = [];
		const regex = new RegExp(this.regex);
		const text = document.getText();
		for (var match of [...text.matchAll(regex)]) {
			try {
				if (typeof match.index === 'undefined') {
					continue;
				}
				const pos = document.positionAt(match.index);
				
				// TODO: Cover by regex
				if(pos.character !== 0){
					continue;
				}

				const range = new Range(pos, pos.translate(undefined, match[0].length));

				if (range) {
					let first = new CodeLens(range);
					(first as unknown as any).order = 1;
					let second = new CodeLens(range);
					(second as unknown as any).order = 2;

					this.codeLenses.push(first);
					this.codeLenses.push(second);
				}
			} catch (e) {
				window.showInformationMessage('Error' + e);
			}
		}
		return this.codeLenses;
	}

	public resolveCodeLens(codeLens: CodeLens | any, token: CancellationToken) {
		const text = window.activeTextEditor?.document.getText(codeLens.range);

		let password = text?.substring(text.indexOf(":") + 1);
		if (!password) {
			return;
		}
		switch (codeLens.order) {
			case 1:
				codeLens.command = {
					title: isEncrypted(password) ? "Unhide Password" : "Hide Password",
					command: "powder.toggle-password-encryption",
					arguments: [codeLens.range],
				};
				break;
			case 2:
				codeLens.command = {
					title: "Copy Password",
					command: "powder.copy-password",
					arguments: [codeLens.range],
				};
				break;
		}

		return codeLens;
	}
}

export async function activate(context: ExtensionContext) {
	let toClean = [];
	const codelensProvider = new CodelensProvider();

	toClean.push(languages.registerCodeLensProvider("*", codelensProvider));

	toClean.push(commands.registerCommand('powder.toggle-password-encryption', async (range) => {
		const text = window.activeTextEditor?.document.getText(range);

		if (!text) {
			return;
		}
		let password = text.substring(text.indexOf(":") + 1);
		if (!password) {
			return;
		}
		const masterPassword = await window.showInputBox({
			placeHolder: "Master password",
			prompt: "Powder: Provide your password to get another",
			value: ''
		});
		if (!masterPassword) {
			return;
		}
		if (isEncrypted(password)) {
			const decryptedPassword = await decrypt(password, masterPassword);
			if (!decryptedPassword) {
				return;
			}
			await window.activeTextEditor?.edit(editBuilder => {
				editBuilder.replace(range.with(new Position(range.start.line, text.indexOf(":") + 1)), decryptedPassword);
			});
		} else {
			const encryptedPassword = await encrypt(password, masterPassword);
			if (!encryptedPassword) {
				return;
			}
			await window.activeTextEditor?.edit(editBuilder => {
				editBuilder.replace(range.with(new Position(range.start.line, text.indexOf(":") + 1)), encryptedPassword);
			});
		}
	}));

	toClean.push(commands.registerCommand('powder.copy-password', async (range) => {
		const text = window.activeTextEditor?.document.getText(range);
		let password = text?.substring(text.indexOf(":") + 1) || null;
		if (!password) {
			return;
		}
		if (isEncrypted(password)) {
			const masterPassword = await window.showInputBox({
				placeHolder: "Master password",
				prompt: "Powder: Provide your password to get another",
				value: ''
			});
			if (!masterPassword) {
				return;
			}
			password = await decrypt(password, masterPassword);
			if (!password) {
				return;
			}
		}
		await env.clipboard.writeText(password);
		window.showInformationMessage('Password has been copied!');
	}));
}

export function deactivate() { }

src/web/base64-utils.ts
/*
MIT License
Copyright (c) 2020 Egor Nepomnyaschih
Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

import { CommentThreadCollapsibleState } from "vscode";

/*
// This constant can also be computed with the following algorithm:
const base64abc = [],
	A = "A".charCodeAt(0),
	a = "a".charCodeAt(0),
	n = "0".charCodeAt(0);
for (let i = 0; i < 26; ++i) {
	base64abc.push(String.fromCharCode(A + i));
}
for (let i = 0; i < 26; ++i) {
	base64abc.push(String.fromCharCode(a + i));
}
for (let i = 0; i < 10; ++i) {
	base64abc.push(String.fromCharCode(n + i));
}
base64abc.push("+");
base64abc.push("/");
*/
const base64abc = [
	"A", "B", "C", "D", "E", "F", "G", "H", "I", "J", "K", "L", "M",
	"N", "O", "P", "Q", "R", "S", "T", "U", "V", "W", "X", "Y", "Z",
	"a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m",
	"n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z",
	"0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "+", "/"
];

/*
// This constant can also be computed with the following algorithm:
const l = 256, base64codes = new Uint8Array(l);
for (let i = 0; i < l; ++i) {
	base64codes[i] = 255; // invalid character
}
base64abc.forEach((char, index) => {
	base64codes[char.charCodeAt(0)] = index;
});
base64codes["=".charCodeAt(0)] = 0; // ignored anyway, so we just need to prevent an error
*/
const base64codes = [
	255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
	255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255,
	255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 62, 255, 255, 255, 63,
	52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 255, 255, 255, 0, 255, 255,
	255, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14,
	15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 255, 255, 255, 255, 255,
	255, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40,
	41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51
];

function getBase64Code(charCode: number) {
	if (charCode >= base64codes.length) {
		throw new Error("Unable to parse base64 string.");
	}
	const code = base64codes[charCode];
	if (code === 255) {
		throw new Error("Unable to parse base64 string.");
	}
	return code;
}

export function bytesToBase64(bytes: Uint8Array) {
	let result = '', i, l = bytes.length;
	for (i = 2; i < l; i += 3) {
		result += base64abc[bytes[i - 2] >> 2];
		result += base64abc[((bytes[i - 2] & 0x03) << 4) | (bytes[i - 1] >> 4)];
		result += base64abc[((bytes[i - 1] & 0x0F) << 2) | (bytes[i] >> 6)];
		result += base64abc[bytes[i] & 0x3F];
	}
	if (i === l + 1) { // 1 octet yet to write
		result += base64abc[bytes[i - 2] >> 2];
		result += base64abc[(bytes[i - 2] & 0x03) << 4];
		result += "==";
	}
	if (i === l) { // 2 octets yet to write
		result += base64abc[bytes[i - 2] >> 2];
		result += base64abc[((bytes[i - 2] & 0x03) << 4) | (bytes[i - 1] >> 4)];
		result += base64abc[(bytes[i - 1] & 0x0F) << 2];
		result += "=";
	}
	return result;
}

export function base64ToBytes(str: string) {
	if (str.length % 4 !== 0) {
		throw new Error("Unable to parse base64 string.");
	}
	const index = str.indexOf("=");
	if (index !== -1 && index < str.length - 2) {
		throw new Error("Unable to parse base64 string.");
	}
	let missingOctets = str.endsWith("==") ? 2 : str.endsWith("=") ? 1 : 0,
		n = str.length,
		result = new Uint8Array(3 * (n / 4)),
		buffer;
	for (let i = 0, j = 0; i < n; i += 4, j += 3) {
		buffer =
			getBase64Code(str.charCodeAt(i)) << 18 |
			getBase64Code(str.charCodeAt(i + 1)) << 12 |
			getBase64Code(str.charCodeAt(i + 2)) << 6 |
			getBase64Code(str.charCodeAt(i + 3));
		result[j] = buffer >> 16;
		result[j + 1] = (buffer >> 8) & 0xFF;
		result[j + 2] = buffer & 0xFF;
	}
	return result.subarray(0, result.length - missingOctets);
}


src/web/crypto-utils.ts
import { bytesToBase64, base64ToBytes } from "./base64-utils";

function getKeyMaterial(password: Uint8Array) {
    return crypto.subtle.importKey(
        "raw",
        password,
        "PBKDF2",
        false,
        ["deriveBits", "deriveKey"]
    );
}

let head = 0;
let version = 1;

export const isEncrypted = (data: string) => {
    try {
        return base64ToBytes(data)[0] === head;
    } catch {
        return false;
    }
};

export const encrypt = async (data: string, password: string) => {
    async function _encrypt(password: Uint8Array, data: Uint8Array, salt: Uint8Array, iv: Uint8Array) {
        let keyMaterial = await getKeyMaterial(password);
        let key = await crypto.subtle.deriveKey(
            {
                "name": "PBKDF2",
                salt: salt,
                "iterations": 100,
                "hash": "SHA-256"
            },
            keyMaterial,
            { "name": "AES-GCM", "length": 256 },
            true,
            ["encrypt", "decrypt"]
        );

        return crypto.subtle.encrypt(
            {
                name: "AES-GCM",
                iv: iv
            },
            key,
            data
        );
    }

    let salt = crypto.getRandomValues(new Uint8Array(12));
    let iv = crypto.getRandomValues(new Uint8Array(12));

    let encrypted = await _encrypt(
        new TextEncoder().encode(password),
        new TextEncoder().encode(data),
        salt,
        iv);
    let packaged = Uint8Array.from([head, version, ...salt, ...iv, ...new Uint8Array(encrypted)]);
    return bytesToBase64(packaged);
};

export const decrypt = async (data: string, password: string) => {
    async function _decrypt(password: Uint8Array, data: Uint8Array, salt: Uint8Array, iv: Uint8Array) {
        let keyMaterial = await getKeyMaterial(password);
        let key = await crypto.subtle.deriveKey(
            {
                "name": "PBKDF2",
                salt: salt,
                "iterations": 100,
                "hash": "SHA-256"
            },
            keyMaterial,
            { "name": "AES-GCM", "length": 256 },
            true,
            ["encrypt", "decrypt"]
        );

        return crypto.subtle.decrypt(
            {
                name: "AES-GCM",
                iv: iv
            },
            key,
            data
        );
    }

    function unpack(encryptedData: string) {
        let data = base64ToBytes(encryptedData);
        if (data[0] !== head || data.length < 12+12+1+1+1) {
            return null;
        }
        return {
            salt: data.subarray(2, 14),
            iv: data.subarray(14, 26),
            encrypted: data.subarray(26, data.length)
        };
    }

    let unpacked = unpack(data);
    if(unpacked === null){
        return null;
    }
    let decrypted = await _decrypt(new TextEncoder().encode(password), unpacked.encrypted, unpacked.salt, unpacked.iv);

    let result = new TextDecoder().decode(decrypted);
    let cat = result.indexOf('\0');
    return result.substring(0, cat !== -1 ? cat : undefined);
};
    